stratum/tools/stratum-replay/stratum_replay.cc

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: Apache-2.0
+#include <iostream>
+#include <memory>
+
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_split.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+#include "grpcpp/grpcpp.h"
+#include "grpcpp/security/credentials.h"
+#include "grpcpp/security/tls_credentials_options.h"
+#include "p4/v1/p4runtime.grpc.pb.h"
+#include "p4/v1/p4runtime.pb.h"
+#include "re2/re2.h"
+#include "stratum/glue/init_google.h"
+#include "stratum/glue/logging.h"
+#include "stratum/glue/status/status.h"
+#include "stratum/glue/status/status_macros.h"
+#include "stratum/hal/lib/p4/utils.h"
+#include "stratum/lib/macros.h"
+#include "stratum/lib/utils.h"
+
+DEFINE_string(grpc_addr, "127.0.0.1:9339", "P4Runtime server address");
+DEFINE_string(p4info, "p4info.pb.txt", "The P4Info file");
+DEFINE_string(pipeline_cfg, "pipeline.pb.bin", "The pipeline config file");
+DEFINE_string(election_id, "0,1",
+              "election id for abstraction update (high,low)");
+DEFINE_uint64(device_id, 1, "The device ID.");
+DEFINE_string(ca_cert, "",
+              "CA certificate, will use insecure credential if empty");
+DEFINE_string(client_cert, "", "Client certificate (optional)");
+DEFINE_string(client_key, "", "Client key (optional)");
+
+namespace stratum {
+namespace tools {
+namespace p4rt_replay {
+
+const char kUsage[] = R"USAGE(
+Usage: stratum-replay [options] [p4runtime write log file]
+  This tool replay P4Runtime write requests to a Stratum device with a given
+  Stratum P4Runtime write request log.
+
+  Options:
+    -device_id: The device ID (default: 1)
+    -election_id: Election ID (high,low) for abstraction update (default: "0,1")
+    -grpc_addr: Stratum gRPC address (default: "127.0.0.1:9339")
+    -p4info: The P4Info file (default: "p4info.pb.txt")
+    -pipeline_cfg: The pipeline config file (default: "pipeline.pb.bin")
+    -ca_cert: CA certificate(optional), will use insecure credential if empty (default: "")
+    -client_cert: Client certificate (optional) (default: "")
+    -client_key: Client key (optional) (default: "")
+)USAGE";
+
+using ClientStreamChannelReaderWriter =
+    ::grpc::ClientReaderWriter<::p4::v1::StreamMessageRequest,
+                               ::p4::v1::StreamMessageResponse>;
+
+::util::Status Main(int argc, char** argv) {
+  if (argc < 2) {
+    LOG(INFO) << kUsage;
+    RETURN_ERROR(ERR_INVALID_PARAM).without_logging() << "";
+  }
+
+  // Initialize the gRPC channel and P4Runtime service stub
+  auto channel_credentials = ::grpc::InsecureChannelCredentials();
+  if (!FLAGS_ca_cert.empty()) {
+    ::grpc::string pem_root_certs;
+    ::grpc_impl::experimental::TlsKeyMaterialsConfig::PemKeyCertPair
+        pem_key_cert_pair;
+    auto key_materials_config =
+        std::make_shared<::grpc_impl::experimental::TlsKeyMaterialsConfig>();
+    ::util::Status status;
+    RETURN_IF_ERROR(
+        ::stratum::ReadFileToString(FLAGS_ca_cert, &pem_root_certs));
```

Comment:
```suggestion
        ReadFileToString(FLAGS_ca_cert, &pem_root_certs));
```

Should work, this is inside the stratum namespace.